Clean without deps: reworking CAN_BE_UNCONFIGURED

gasche opened the pull request:
This PR is a non-trivial rework of the CAN_BE_UNCONFIGURED logic to fix a build performance regression caused by #9332: in the current trunk, `make clean` starts by recomputing the C dependencies, adding about 8 seconds of useless work.

(If you wonder what `CAN_BE_UNCONFIGURED` is: it is the list of the current Makefile targets which *require* `./configure` to have been executed first (this excludes `*clean` and `configure`). Yes, this doesn't make complete sense, and this PR renames it to `REQUIRES_CONFIGURATION`. The point is that our Makefiles check if the list is empty, and then go into warrior mode to work correctly without configuration.)

We fix this in two daring steps:
1. Move the inclusion of `Makefile.build_config` into `Makefile.common`
    (This simplifies all makefiles slightly: the two files were always included in this order.)
2. When `CAN_BE_UNCONFIGURED` is empty, do *not* include `Makefile.build_config`
    (previously we would `-include`, which would tolerate the absence without failing)

The risky part is the second one, and it needs a careful review -- but it should be safe as we supported not having the file at all before, right?

As a result of the second change, the `.dep` logic is not included when there are no REQUIRES_CONFIGURATION target, and in particular `make clean` does not try to compute C dependencies anymore.

On my machine, `make distclean` goes from 8.5s to 0.5s with this PR.

(cc @dra27 @shindere)

gasche committed:
makefiles: move the inclusion of Makefile.build_config in Makefile.common

gasche committed:
[minor] rename CAN_BE_UNCONFIGURED into REQUIRES_CONFIGURATION

gasche committed:
build system: skip Makefile.build_config when REQUIRES_CONFIGURATION is empty

Before we would include it if present, but support building without
it. Unless I misunderstand something, this is unnecessarily complex:

- If we want to support building without it, we may as well
  do without it all the time (in particular it makes it easier
  to spot bugs which occurs when it is missing).

- Certain makefile rules (see config.status) work by being added
  as dependencies to Makefile.build_config, and requiring this file
  only when actually needed makes it possible to simplify those.
  (simplifications not included in this commit)

I reviewed all 'clean' rules to check that they do not depend on
variables defined in the configure. On the other hand, the main
makefiles do a lot of conditional tests on those variables,
so --warn-undefined-variables warns quite a bit. (This could be solved
by a Makefile.no_config rule that mirrors Makefile.config with dummy
definitions.)

gasche committed:
Makefile: no need for REQUIRES_CONFIGURATION check on config.status

Now that Makefile.build_config is only included when configuration is
required, we can remove the REQUIRES_CONFIGURATION logic here.

dra27 commented:
The rationale behind `CAN_BE_UNCONFIGURED` is that it's a boolean, which in `make` means a non-empty string or the empty string. That said, it's the wrong way around for that which I agree was a tad weird - the rename to `REQUIRES_CONFIGURATION` is more logical.

I have gone through the `Makefile`s carefully and I agree with the changes. Checking the clean targets again has thrown up a few things which could profitably be fixed here if you wish:
- in the root `Makefile`, `ARCH_SPECIFIC` comes from `compilerlibs/Makefile.compilerlibs` and it's influenced by `ARCH`. For the clean target, that should be reworked not to require `ARCH`. I expect the way to do it is to remove the relevant `clean::` target in `Makefile` completely and have smaller fragments in `compilerlibs/Makefile.compilers` which unconditionally remove each architecture's specific files.
- `ocamltest/Makefile` has a very carefully constructed `clean` target which tries to clean exactly what's built. As a result, it depends on `UNIX_OR_WIN32` (with a manual case to remove the other files). It's not particularly critical, but I checked other uses and perhaps the simplest thing is in `Makefile.common` to have `UNIX_OR_WIN32=unix` for the case where `Makefile.build_config` is _not_ included.
- in `stdlib/Makefile` there's an `rm -f camlheader*` which is redundant (there's an `rm -f $(CAMLHEADERS)` above which not only includes that pattern but is also responsible for removing the `target_*` files)
- `testsuite/Makefile` includes `Makefile.tools` which has a `-include Makefile.config` in it. It's harmless, but is the only remaining place where `REQUIRES_CONFIGURATION` could be used. That said, IIRC, the testsuite doesn't use `Makefile.common`.

gasche commented:
This morning I tried to create a Makefile.no_config file that ensures that no "undefined variable" warning is shown when running one of the no-configure targets. This is the patch I ended up with:

```diff
diff --git a/Makefile.common b/Makefile.common
index 45cf7bff26..f84453b460 100644
--- a/Makefile.common
+++ b/Makefile.common
@@ -28,6 +28,8 @@ REQUIRES_CONFIGURATION := $(strip \
 
 ifneq "$(REQUIRES_CONFIGURATION)" ""
 include $(ROOTDIR)/Makefile.build_config
+else
+include $(ROOTDIR)/Makefile.no_config
 endif
 
 # %(DEPDIR) must be kept in sync with entries in .gitignore
diff --git a/Makefile.no_config b/Makefile.no_config
new file mode 100644
index 0000000000..6717501a90
--- /dev/null
+++ b/Makefile.no_config
@@ -0,0 +1,46 @@
+# from Makefile.config
+EMPTY=
+
+O=o
+A=a
+EXE=exe
+SO=so
+
+UNIXLIB=unix
+UNIX_OR_WIN32=
+FUNCTION_SECTIONS=
+FLAMBDA=
+
+NATIVE_COMPILER=
+NATDYNLINKOPTS=
+CCOMPTYPE=
+ARCH=
+SYSTEM=
+WITH_OCAMLDOC=
+WITH_DEBUGGER=
+STDLIB_MANPAGES=
+TOOLCHAIN=
+NATDYNLINK=
+WITH_CAMLTEX=
+
+TARGET=
+HOST=
+
+INSTALL_BYTECODE_PROGRAMS=
+INSTALL_SOURCE_ARTIFACTS=
+MKEXE=
+MKDLL=
+
+FLEXLINK=
+
+ARCH64=
+RUNTIMED=
+RUNTIMEI=
+SUPPORTS_SHARED_LIBRARIES=
+
+SHEBANGSCRIPTS=
+HASHBANGSCRIPTS=
+OTHERLIBRARIES=
+
+# From Makefile.build_config
+COMPUTE_DEPS=false
diff --git a/Makefile.tools b/Makefile.tools
index 49f4d2f67f..67ba7134fc 100644
--- a/Makefile.tools
+++ b/Makefile.tools
@@ -47,7 +47,16 @@ SET_LD_PATH=CAML_LD_LIBRARY_PATH="$(LD_PATH)"
 #   variable. Note that for Windows we add Unix-syntax directory names in
 #   PATH, and Cygwin will translate it to Windows syntax.
 
--include $(TOPDIR)/Makefile.config
+# See Makefile.common
+REQUIRES_CONFIGURATION := $(strip \
+  $(filter-out partialclean clean distclean configure, \
+	$(if $(MAKECMDGOALS),$(MAKECMDGOALS),defaultentry)))
+ifneq "$(REQUIRES_CONFIGURATION)" ""
+include $(TOPDIR)/Makefile.config
+else
+include $(TOPDIR)/Makefile.no_config
+endif
+
 
 # Make sure USE_RUNTIME is defined
 USE_RUNTIME ?=
```

I have the impression that some of your comments are going in the same direction (in particular re Makefile.tools). I don't really like my patch (it feels wrong to define a bunch of variables to an invalid state to silence a warning), but I may try to salvage some parts according to your comment, and you might have en enlightening comment to make about it.

dra27 commented:
The problem is that the undefined variable warning is emitting so many false positives in that case it's not helpful - you really want a warning from `make` only if the variable might have made a semantic difference (which would substantially reduce the number of those definitions).

There is at least a CI test that `make distclean` cleans things properly now - perhaps that's sufficient. The problem is that if a new variable is introduced in `Makefile.config.in` it has to go in another file.

Maybe the better approach is just to ensure that ocamltest's `Makefile` really doesn't depend on `$(UNIX_OR_WIN32`) (I think that might just mean defining a slightly different sources variable which doesn't use `$(run)`)?

gasche commented:
> The problem is that if a new variable is introduced in Makefile.config.in it has to go in another file.

I also thought about defining each variable in a REQUIRES_CONFIGURATION conditional in Makefile.config, so that the "default" value is given at the same time, but I think that it is worse.

> Maybe the better approach is just to ensure that ocamltest's Makefile really doesn't depend on $(UNIX_OR_WIN32).

Yes, we should do this in any case. It's not nice to depend on configuration too much, because if I reconfigure and then clean it may miss certain artifacts of my previous build.

gasche committed:
move the REQUIRES_CONFIGURATION logic in Makefile.config_if_required

This lets us share the logic between Makefile.common and Makefile.tools.

gasche committed:
refactor Makefile.config_if_required for easier reuse of variables

gasche committed:
ocamltest/Makefile: do not depend on UNIX_OR_WIN32 for 'clean'

Change suggested by David Allsopp.

gasche commented:
@dra27 I pushed three further commits that address Makefile.tools and the ocamltest clean rule.

I had misunderstood the situation of the ocamltest clean rule in my comments above; in particular, it does already (in trunk) clean the artifacts corresponding to all possible configurations. The dependency on UNIX_OR_WIN32 is not important to its behavior (given that it has a line to explicitly clean all `run_*.o` artifacts anyway), but it is a result of the nice factorization of the logic to decide the source files of a particular configured run.

It took me a few iterations to find an approach that I thought was reasonable, given the neatness of the general makefile.

gasche committed:
[minor] Makefile: move ARCH_SPECIFIC back to the main Makefile

(This is a small, independent readability fix.)

ARCH_SPECIFIC was moved into Makefile.compilerlibs by mistake; it is
not used in Makefile.compilerlibs (in particular the
ARCH_SPECIFIC_ASMCOMP_* variables do not depend on it), and it is used
in the main Makefile.

gasche committed:
[minor] stdlib/Makefile: remove redundant clean action

There is already a more precise

    clean::
            rm -f $(CAMLHEADERS)

rule above.

(Change suggested by David Allsopp.)

gasche commented:
@dra27 I went over your two remaining suggestions:

- I fixed `stdlib/Makefile` as you suggested.
- In fact I believe there is no problem with `ARCH_SPECIFIC` except a mistake in Makefile structure which made the code confusing. I moved it back to the main Makefile. Its value does *not* depend on `ARCH`.

Would you be willing to do a review of the PR?

gasche committed:
Changes entry

dra27 commented:
Nice change to ocamltest's `Makefile`! Agree with the rest (including my confusion with `$(ARCH_SPECIFIC)` - the source files are of course not cleaned, and those files don't created specially for `clean`)

The `Makefile.config_if_required` uses `ROOTDIR` but the manual and testsuite, for legacy reasons, use `TOPDIR`. Up to you if you want to grasp the bull by the horns and change the testsuite and manual `Makefile`s to use `ROOTDIR` like the rest of the build system (I expect this is on @shindere's hit list for some day, as he made the use of `ROOTDIR` uniform in the main build system) or, more simply, either have `ROOTDIR ?= $(TOPDIR)` in `Makefile.config_if_required` (that will probably upset the undefined variables test, though) or put `ROOTDIR=$(TOPDIR)` in `testsuite/Makefile` before `Makefile.tools` gets included

gasche commented:
Fun story time: my internet box seemingly died this morning, not the best time. I now use 3G from some fallback simcard, and emails and github are *okay* but for some reason I cannot manage to access Travis/Appveyors results.

(I think that your message explains what the failure is, thanks for that.)

I will define ROOTDIR in the TOPDIR-abiding makefiles. We should migrate those to use Makefile.common, but I'm not up to the challenge for now. I also realize from your comment that I probably completely forgot to look at the manual Makefile, will do now.

I'm starting to get tired of the `Makefile.foo` files in the root directory. I would like to move most of them (all of them except `Makefile`, `Makefile.config` and `Makefile.build_config`) into a `make/` subdirectory, but I will propose this as a separate PR.

dra27 commented:
Hah - it's obviously contagious! My main broadband provider is having a bad day... after much head-scratching, I discovered what had gone wrong with the configuration for the second broadband provider, but switching to that has now cut me off from our lab cluster, because I'm now on a different router 🤯 The CI failure across the board is that the testsuite tries to read `/Makefile.build_config`, yes.

gasche commented:
(Now Travis is happy. Is David also happy?)

dra27 commented:
Please could you rebase it on to trunk (Xavier just merged the C dependencies fixes) and then we do a run of `precheck` and let's merge?

dra27 commented:
(I am certainly happy, even if poor Sébastien's post-holiday-catch-up-review spike seems to be ever-growing!!)

gasche commented:
I just rebased against trunk.

shindere commented:
While working on #9652, I noticed that `make clean` & co do not work
very well if one lets them run without loading the configuration files.

The clean targets could probably be enhanced so that they do a proper
job even on an unconfigured tree, but it felt simpler to me just to
_not_ ignore the configuration files, if they exist.

That's why #9652 includes a commit that uses `-include` rather than
`include` to load configuration files for clean targets, so that
the configuration gets included if it exists and ignored if it does not.

gasche commented:
@shindere I'm nervous about the change you propose:

- The present PR fixes a performance regression in `make (dist)clean`; can you confirm that your change avoids reintroducing the regression?
- In the current trunk, there are two possible states to keep in mind: either REQUIRES_CONFIGURATION is empty, and the configuration is not loaded, or it is not, and the configuration is loaded. The change you propose introduces three configurations: either REQUIRES_CONFIGURATION is non-empty and the configuration is loaded, or it is empty and the configuration is or is not loaded. I am worried that we will introduce build system bugs because of this extra complexity.

(Not loading the configuration results in an unpleasant programming model for the build system, but the fact that we have to deal with this predates the current PR, and I don't see how we could avoid it.)

I would rather be in favor of discussing and fixing the `clean` targets in absence of configuration, if this can be done reasonably. At least exploring the option.

shindere commented:
Gabriel Scherer (2020/06/09 01:46 -0700):
> @shindere I'm nervous about the change you propose:
>
> - The present PR fixes a performance regression in `make (dist)clean`;
> can you confirm that your change avoids reintroducing the regression?

Yes

damiendoligez commented:
> I would rather be in favor of discussing and fixing the `clean` targets in absence of configuration, if this can be done reasonably. At least exploring the option.

The proper solution would be to clean all **possible** targets but that requires iterating on all possible values of `$(EXE)` and maybe other variables.

damiendoligez commented:
Another proper solution would be to separate the configuration from the computation of the C dependencies

gasche commented:
Yes, we can have a mode where the configuration is loaded, but the dependencies are not computed (this was my initial approach when starting the work that became this PR, only I realize on the way that the present approach also worked and seemed simpler).

We could avoid the "three states" problem I mentioned above by ensuring that, if configuration has not run, `clean` targets are obvious no-ops (possibly printing a message). Then we are back in the two-state situation. This should be discussed with @dra27 would introduced the do-something-without-configuration logic, if I understand correctly.

Note that several of our clean targets already use the "proper solution". For example, `ocamltest/Makefile`'s clean rule will remove both `ocamltest` and `ocamltest.exe`, no matter what the configuration is; same with `runtime/Makefile`. (In theory this approach is more robust to artifacts coming from an older, different configuration, although in practice I am not sure it matters that much.)

shindere commented:
Damien Doligez (2020/06/09 02:59 -0700):
> > I would rather be in favor of discussing and fixing the `clean` targets in absence of configuration, if this can be done reasonably. At least exploring the option.
>
> The proper solution would be to clean all **possible** targets but
> that requires iterating on all possible values of `$(EXE)` and maybe
> other variables.

That's it: making `clean` general enough to remove all the things that
could be generated will become harder and harder and at the same time
somehow less generic: if one day $(EXE) happens to take another value
than those it currently has, we will have to rework our clena rules. The
example here with $(EXE) might not be very convincing, but it is
important ot see it's just an example andwhat happens here can also
happen for other variables. Not only that it can happen, it happens.

shindere commented:
Damien Doligez (2020/06/09 03:01 -0700):
> Another proper solution would be to separate the configuration from
> the computation of the C dependencies

Well it is separated, it's jsut that we need to avoid the dependencies
to get recomputed. I think that what would have been achieved if we used
`-include` rather than `include` to load the `.d` files, but as far as I
can remember @dra27 didn't like this solution for a reason that I am not
able to remember right now.

shindere commented:
Gabriel Scherer (2020/06/09 10:07 +0000):
> Yes, we can have a mode where the configuration is loaded, but the dependencies are not computed (this was my initial approach when starting the work that became this PR, only I realize on the way that the present approach also worked and seemed simpler).
>
> We could avoid the "three states" problem I mentioned above by ensuring that, if configuration has not run, `clean` targets are obvious no-ops (possibly printing a message). Then we are back in the two-state situation. This should be discussed with @dra27 would introduced the do-something-without-configuration logic, if I understand correctly.
>
> Note that several of our clean targets already use the "proper
> solution". For example, `ocamltest/Makefile`'s clean rule will remove
> both `ocamltest` and `ocamltest.exe`, no matter what the configuration
> is; same with `runtime/Makefile`. (In theory this approach is more
> robust to artifacts coming from an older, different configuration,
> although in practice I am not sure it matters that much.)

As @damiendoligez said, being completely general looks harder than it
may seem by considering just the $(EXe) case.

gasche commented:
I don't personally care about being "completely general" as long as it works for the use-cases we care about. There is no clear specification of what `clean` should do (should it actually remove artifacts coming from a previous-and-different configuration?). That being said, you should of course feel free to propose better approaches than the one in this PR, as long as they do not require computing the dependencies to clean.

shindere commented:
Gabriel Scherer (2020/06/09 03:26 -0700):
> I don't personally care about being "completely general" as long as it
> works for the use-cases we care about. There is no clear specification
> of what `clean` should do (should it actually remove artifacts coming
> from a precious-and-different configuration?). That being said, you
> should of course feel free to propose better approaches than the one
> in this PR, as long as they do not require computing the dependencies
> to clean.

I confirm they odn't and agree that it would be a bug and it was fully
legitimate to fix it.

dra27 commented:
Since #8639, `make clean` and `make distclean` should clean everything even when the tree isn’t configured. What’s not working with that, as it’s supposed to be checked by Travis as well?

gasche commented:
Thinking about this slightly more, I think there are two sensible approaches:

1. Ensure that our makefile works reasonably without configuration (for rules that do not require building new artifacts), and use this usage mode for `clean`, trying to clean all artifacts of supported configurations.

2. Ensure that we never work without configuration (either we fail, for most rules, or we return immediately with a message, possibly for clean rules), and require a configuration to do any cleaning work.

The advantages of (1) is that it is currently implemented, and that @dra27 seems in favor of it. The main disadvantage is that it is unpleasant to reason about the build system code in absence of configuration. (We have plenty of undefined variables, which is maybe fine, but one is not used to making a distinction between `ifeq "$(FOO)" "true"` and `ifneq "$(FOO)" "false"`.)

If one wanted to go for (2), we would need another distinction: scenarios where we have a configuration, but we don't want to compute dependencies (typically for cleaning, but maybe other rules as well?). Supporting this well/robustly may require some care in the build system. (Ideally it should be engineered so that it is easy to ensure that no dependency-computation happens for certain rules, just like it is currently easy to see that dependency computation only happens if configuration is required.)

shindere commented:
I'll do my best to think about all this and find documentation if
possible.

Meanwhile, I'd really appreciate it if #9652 could be reviewed (the
point we discuss here is really a small part of it).

It would be nice if our discussion here would not be a blocker for #9652
because that PR does fix an issue. So, I hope that the fact it does not
introduce a regression will be enough for it to be accepted.

shindere commented:
David Allsopp (2020/06/09 11:16 +0000):
> Since #8639, `make clean` and `make distclean` should clean everything
> even when the tree isn’t configured.

Am I correct that this is kind of just a matter of taste, and at least
that it is not mandated by any standard or even convention that would be
documented somewhere?

Not to say we shouldn't do it, just ot understand where it comes from.

shindere commented:
I'd definitely go for 2 but understand it's not going to happen.

I think it's not ucommon in the open-osurce world that the root Makefile
is actually obtained from a Makefile.in template file, meaning that you
simply can't run make as long as theproject has not been configured. I
agree that in such a situation there is no message telling what to do
but feel it is completely okay and would keep things really simple.

And yes of course care should be taken to compute dependencies only when
they are strictly needed. It would feel more accurate to me to not
compute dependencies when cleaning than to not load configuration.

gasche commented:
Well that's something for build-system maintainers (I would think of @shindere, @dra27 and possibly @damiendoligez) to discuss and build a consensus on. I'm happy to let other people make decisions on these questions.

shindere commented:
Well now that you have done this PR, you are involved, too! ;-)

gasche commented:
I rationalize my choice to not be involved too much by saying that this PR is mostly a cleanup/explicitation of the design choice (1) already made in trunk at the time, rather than a decision on my side.
